Add server support for stage2, stage3, and report targets

stephen-soltesz opened the pull request:
This change adds new targets to the epoxy server for handling stage2, stage3, and report.

This change includes significant revisions to the storage.Host type and associated tests. These name updates are needed to support a three stage boot sequence (previously two were planned).

Additionally, this change removes the propagation of errors due to failed reads from `rand.Reader`. These reads are extremely unlikely in normal environments, and even in those cases, the panic will cause the HTTP request to fail.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/epoxy/27)
<!-- Reviewable:end -->

pboothe commented:
Reviewed 1 of 8 files at r1.
Review status: 1 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cmd/epoxy_boot_server/main.go, line 105 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4frsjT4kB947LtTYPX:-L4frsjUqgyTXQ0LFZlG:b4drks9) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/cmd/epoxy_boot_server/main.go#L105)):*
> ```Go
> 
> 	// Stage2, stage3, and report targets load after stage1 runs successfully.
> 	addRoute(router, "POST", "/v1/boot/{hostname}/{sessionId}/stage2",
> ```

Add `//TODO: make the names stage2 and stage3 better when we need to support the case where not every machine has the same stage2 or stage3`

---

*[storage/datastore_test.go, line 92 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4fy1LElY6QbBWmbP7L:-L4fy1LElY6QbBWmbP7M:bbll10l) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/storage/datastore_test.go#L92)):*
> ```Go
> 		},
> 		CurrentSessionIDs: SessionIDs{
> 			Stage2ID: "01234",
> ```

This "combine the stage name as a prefix for the thing you want" is worrying.  I think I would prefer if the stage name were the complete key for the data. Then SessionIDs would just be a map like the Sequence, and their keys would be exactly the same instead of doing this prefix thing.  This is an optional request, though, as the code is out of scope for this PR.

---

*[storage/host.go, line 54 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4fwjS5Uz5WUYPBuCjE:-L4fwjS5Uz5WUYPBuCjF:b-4ul006) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/storage/host.go#L54)):*
> ```Go
> type SessionIDs struct {
> 	Stage2ID string // Needed for requesting the nextstage.json target.
> 	Stage3ID string // Needed for requesting the begingstage target.
> ```

beginstage

---

*[storage/host.go, line 59 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4fwpKidLA7DH4-6EFW:-L4fwpKidLA7DH4-6EFX:b-9ynkrx) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/storage/host.go#L59)):*
> ```Go
> 
> // Sequence represents a set of operator-provided iPXE scripts or JSON nextboot Configs.
> type Sequence struct {
> ```

Should be a map from string->string. "stage1" or "Stage1ChainURL" can then be keys and the URLs values. This approach would already make things less kludgy - you currently have a test where Stage3ChainURL  is set to "", when really it should never be set at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/27)*
<!-- Sent from Reviewable.io -->

pboothe commented:
All of my change requests only need to be fixed now if they affect the data served up to a client. Otherwise, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/27#-:-L4fz7FVyWNi3BmWvLdX:bqtvxkt)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thank you for the thoughtful review!

Today and in the future there will have to be agreement between the client and server regarding the name of the target in the URL, e.g. "stage2" or "stage3", "report". But the content of the response is only what's defined in nextboot.Config.

The Host definition is wholly contained within Datastore and should be changeable without impacting the actual server.

I'd like to submit this PR as is, since I believe we share an understanding about the preferred shape of the Host types.

---

Review status: 1 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cmd/epoxy_boot_server/main.go, line 105 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4frsjT4kB947LtTYPX:-L4g7pIpW_T4yc1yLwe5:bfg6brm) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/cmd/epoxy_boot_server/main.go#L105)):*
> //TODO: make the names stage2 and stage3 better when we need to support the case where not every machine has the same stage2 or stage3
 
Added TODO: s/better/arbitrary/

---

*[storage/datastore_test.go, line 92 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4fy1LElY6QbBWmbP7L:-L4g9gr4k_2K6bCT6p8u:b-fvtafd) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/storage/datastore_test.go#L92)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

This "combine the stage name as a prefix for the thing you want" is worrying.  I think I would prefer if the stage name were the complete key for the data. Then SessionIDs would just be a map like the Sequence, and their keys would be exactly the same instead of doing this prefix thing.  This is an optional request, though, as the code is out of scope for this PR.
</blockquote></details>

I was feeling like there was something wrong and I think you've named it.

I've added a TODO in hosts.go for the Sequence & SessionIDs types. In my mind, making the target names arbitrary plus the changes to Sequence & SessionIDs are all connected.

---

*[storage/host.go, line 54 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4fwjS5Uz5WUYPBuCjE:-L4g962wRx4xrxaOxH8P:b-1jl00q) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/storage/host.go#L54)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

beginstage
</blockquote></details>

Fixed. Thank you.

---

*[storage/host.go, line 59 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/27#-L4fwpKidLA7DH4-6EFW:-L4gBMRC0xpGAviBNzO9:btd51a6) ([raw file](https://github.com/m-lab/epoxy/blob/0daf2df7cb9823fe74d70d5f2ac897fcf98342d5/storage/host.go#L59)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Should be a map from string->string. "stage1" or "Stage1ChainURL" can then be keys and the URLs values. This approach would already make things less kludgy - you currently have a test where Stage3ChainURL  is set to "", when really it should never be set at all.
</blockquote></details>

I agree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/27#-:-L4gBWN0Yk-5lyP__d2a:bdrj78t)*
<!-- Sent from Reviewable.io -->
